tempesta_fw/apm.c

```diff
@@ -0,0 +1,1041 @@
+	percentile->seq = rdidx;
+
+	return (seq != rdidx);
+}
+
+/*
+ * Verify that an APM Stats user using the same set of percentiles.
+ *
+ * Note: This module uses a single set of percentiles for all servers.
+ * All APM Stats users must use the same set of percentiles.
+ */
+int
+tfw_apm_percentile_verify(Percentile *percentile)
+{
+	int i;
+
+	if (percentile->pstsz != ARRAY_SIZE(tfw_apm_pstats))
+		return 1;
+	for (i = 0; i < percentile->pstsz; ++i)
+		if (percentile->pstats[i].ith != tfw_apm_pstats[i].ith)
+			return 1;
+	return 0;
+}
+
+static inline void
+__tfw_apm_update(TfwApmRBuf *rbuf, unsigned long jtstamp, unsigned long jrtime)
+{
+	int centry = (jtstamp / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+	unsigned long jtmistart = jtstamp - (jtstamp % tfw_apm_jtmintrvl);
+	TfwApmRBEnt *crbent = &rbuf->rbent[centry];
+
+	tfw_apm_rbent_checkreset(crbent, jtmistart);
+	tfw_stats_update(&crbent->pcntrng, jiffies_to_msecs(jrtime));
+}
+
+void
+tfw_apm_update(void *apmdata, unsigned long jtstamp, unsigned long jrtime)
+{
+	BUG_ON(!apmdata);
+	__tfw_apm_update(&((TfwApmData *)apmdata)->rbuf, jtstamp, jrtime);
+}
+
+/*
+ * Destroy the specified APM ring buffer.
+ */
+void
+tfw_apm_destroy(void *apmdata)
+{
+	TfwApmData *data = apmdata;
+
+	clear_bit(TFW_APM_DATA_F_REARM, &data->flags);
+	smp_mb__after_atomic();
+	del_timer_sync(&data->timer);
+
+	kfree(data);
+}
+
+/*
+ * Initialize a ring buffer entry.
+ */
+static inline void
+tfw_apm_rbent_init(TfwApmRBEnt *rbent, unsigned long jtmistamp)
+{
+	memcpy(rbent->pcntrng.ctl, tfw_rngctl_init, sizeof(rbent->pcntrng.ctl));
+	__tfw_apm_rbent_reset(rbent, jtmistamp);
+}
+
+/*
+ * Create and initialize an APM ring buffer for a server.
+ */
+void *
+tfw_apm_create(void)
+{
+	TfwApmData *data;
+	TfwApmRBEnt *rbent;
+	PrcntlStats *pstats[2];
+	int i, size;
+	int rbufsz = tfw_apm_tmwscale;
+	int pstsz = ARRAY_SIZE(tfw_apm_pstats);
+
+	if (!tfw_apm_tmwscale) {
+		TFW_ERR("Late unitialization of `apm_stats` option\n");
+		return NULL;
+	}
+
+	/* Keep complete stats for the full time window. */
+	size = sizeof(TfwApmData) + rbufsz * sizeof(TfwApmRBEnt)
+				  + 2 * sizeof(tfw_apm_pstats);
+	if ((data = kzalloc(size, GFP_ATOMIC)) == NULL)
+		return NULL;
+
+	rbent = (TfwApmRBEnt *)(data + 1);
+	pstats[0] = (PrcntlStats *)(rbent + rbufsz);
+	pstats[1] = (PrcntlStats *)(pstats[0] + pstsz);
+
+	data->rbuf.rbent = rbent;
+	data->rbuf.rbufsz = rbufsz;
+
+	data->stats.prcntl[0].pstats = pstats[0];
+	data->stats.prcntl[0].pstsz = pstsz;
+
+	data->stats.prcntl[1].pstats = pstats[1];
+	data->stats.prcntl[1].pstsz = pstsz;
+
+	for (i = 0; i < rbufsz; ++i)
+		tfw_apm_rbent_init(&rbent[i], 0);
+
+	memcpy(pstats[0], tfw_apm_pstats, sizeof(tfw_apm_pstats));
+	memcpy(pstats[1], tfw_apm_pstats, sizeof(tfw_apm_pstats));
+
+	rwlock_init(&data->stats.prcntl[0].rwlock);
+	rwlock_init(&data->stats.prcntl[1].rwlock);
+	atomic_set(&data->stats.rdidx, 0);
+
+	set_bit(TFW_APM_DATA_F_REARM, &data->flags);
+	smp_mb__after_atomic();
```

Comment:
`setup_timer()` is a function call from external module - i don't believe compiler can replace the `set_bit` operation with the function call.